components/script/dom/videotracklist.rs

```diff
@@ -0,0 +1,172 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::VideoTrackListBinding::{self, VideoTrackListMethods};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::refcounted::Trusted;
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject};
+use crate::dom::bindings::root::Dom;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::videotrack::VideoTrack;
+use crate::dom::window::Window;
+use crate::task_source::TaskSource;
+use dom_struct::dom_struct;
+
+#[dom_struct]
+pub struct VideoTrackList {
+    eventtarget: EventTarget,
+    tracks: DomRefCell<Vec<Dom<VideoTrack>>>,
+}
+
+impl VideoTrackList {
+    pub fn new_inherited(tracks: &[&VideoTrack]) -> VideoTrackList {
+        VideoTrackList {
+            eventtarget: EventTarget::new_inherited(),
+            tracks: DomRefCell::new(tracks.iter().map(|track| Dom::from_ref(&**track)).collect()),
+        }
+    }
+
+    pub fn new(window: &Window, tracks: &[&VideoTrack]) -> DomRoot<VideoTrackList> {
+        reflect_dom_object(
+            Box::new(VideoTrackList::new_inherited(tracks)),
+            window,
+            VideoTrackListBinding::Wrap,
+        )
+    }
+
+    pub fn len(&self) -> usize {
+        self.tracks.borrow().len()
+    }
+
+    pub fn item(&self, idx: usize) -> Option<DomRoot<VideoTrack>> {
+        self.tracks
+            .borrow()
+            .get(idx)
+            .map(|track| DomRoot::from_ref(&**track))
+    }
+
+    pub fn selected_index(&self) -> Option<usize> {
+        self.tracks
+            .borrow()
+            .iter()
+            .position(|track| track.selected())
+    }
+
+    // TODO(#22799) Integrate DOM Audio and Video track selection with media player
+    pub fn set_selected(&self, idx: usize, value: bool) {
```

Comment:
Hopefully I got the logic correct here on selecting the video track. I am thinking it can be cleaned up more though.